File: algoritmos/recursao/java/ED2Lista1_questao05.java
/**
 * Função : Inverter uma string usando recursão.
 * Autor  : José Carlos da Silva Júnior (UNIPÊ)
 * Data   : 01/10/2025
 * Observações: Q5 - Recursão.
 */
package algoritmos.recursao.java;
import java.util.Scanner;

public class ED2Lista1_questao05 {
    
    // Função recursiva para inverter a string
    public static String inverterString(String str) {
        // Caso base: se a string estiver vazia ou tiver um caractere, retorna a própria string
        if (str.isEmpty()) {
            return str;
        }
        // Chamada recursiva: pega o último caractere e concatena com a inversão do restante da string
        return str.charAt(str.length() - 1) + inverterString(str.substring(0, str.length() - 1));
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        // Solicita ao usuário que insira uma string
        System.out.print("Digite uma string para inverter: ");
        String inputString = scanner.nextLine();
        
        // Chama a função recursiva para inverter a string
        String invertedString = inverterString(inputString);
        
        // Exibe a string invertida
        System.out.println("String invertida: " + invertedString);
        
        scanner.close();
    }
}

File: algoritmos/recursao/c/ED2Lista1_questao05.c
/**
 * Função : Inverter uma string usando recursão.
 * Autor  : José Carlos da Silva Júnior (UNIPÊ)
 * Data   : 01/10/2025
 * Observações: Q5 - Recursão.
 */

#include <stdio.h>
#include <stdlib.h>

void inverter_string(char *str, int inicio, int fim) {
    if (inicio >= fim) {
        return;
    }
    
    // Troca os caracteres nas posições 'inicio' e 'fim'
    char temp = str[inicio]

    str[inicio] = str[fim];
    str[fim] = temp;
    
    // Chama a função recursivamente para o próximo par de caracteres
    inverter_string(str, inicio + 1, fim - 1);
}

int main() {
    char str[] = "Estrutura de dados"; // Exemplo de string
    int len = 0;
    
    // Calcula o comprimento da string
    while (str[len] != '\0') {
        len++;
    }
    
    inverter_string(str, 0, len - 1);
    
    printf("String invertida: %s\n", str);
    
    return 0;
}